### PR TITLE
[WLANSVC] Improve _RpcEnumInterfaces

### PR DESCRIPTION
**Pull request by christophvw**
## Purpose

implement _RpcEnumInterfaces without relying on Vista+ functions

JIRA issue: [CORE-12275](https://jira.reactos.org/browse/CORE-12275)

## TODO
- [ ]  remove breakpoint
- [ ]  remove debug prints
- [ ]  fix unmarshall crash at function return

### CODE REVIEW

**Comment by binarymaster** [comment]
You might want to mention ticket id (CORE-12275) in the commit message, it will associate the commit with the JIRA issue.

File: base/services/wlansvc/rpcserver.c [hunk 1]
```diff
@@ -171,8 +172,87 @@ DWORD _RpcEnumInterfaces(
...
+    {
+        HeapFree(GetProcessHeap(), 0, pIfTable);
+        pIfTable = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, dwSize);
+        if (!pIfTable)
+        {
+            return ERROR_NOT_ENOUGH_MEMORY;
+        }
+    }
+
+    if ((GetIfTable(pIfTable, &dwSize, FALSE)) == NO_ERROR)
```
Comment (HBelusca) [inline base/services/wlansvc/rpcserver.c]: Remove double parenthesizing.

File: base/services/wlansvc/rpcserver.c [hunk 2]
```diff
@@ -171,8 +172,87 @@ DWORD _RpcEnumInterfaces(
...
+    {
+        dwSize = (pIfTable->dwNumEntries * sizeof(WLAN_INTERFACE_INFO));
+
+        DPRINT1("NumEntries %ld\n",pIfTable->dwNumEntries);
+        __debugbreak();
```
Comment (HBelusca) [inline base/services/wlansvc/rpcserver.c]: Debug leftover?

File: base/services/wlansvc/rpcserver.c [hunk 3]
```diff
@@ -171,8 +172,87 @@ DWORD _RpcEnumInterfaces(
...
+        *ppInterfaceList = InterfaceList;
+
+        if (!pIfTable->dwNumEntries)
+        {
+            return ERROR_SUCCESS;
+        }
+
+        for (i = 0; i < pIfTable->dwNumEntries; i++)
+        {
+
```
Comment (HBelusca) [inline base/services/wlansvc/rpcserver.c]: extra newline

File: modules/rostests/apitests/wlanapi/wlanapi.c [hunk 4]
```diff
@@ -188,15 +188,36 @@ static void WlanGetProfile_test(void)
...
+    if (ret == ERROR_SERVICE_EXISTS)
+    {
+        skip("Skipping tests, WlanSvc is not running\n");
+        return;
+    }
+
+    ret = WlanEnumInterfaces(hClientHandle, NULL, &pInterfaceList);
+    ok(ret == ERROR_SUCCESS, "expected success\n");
+    
+
```
Comment (HBelusca) [inline modules/rostests/apitests/wlanapi/wlanapi.c]: extra newline

File: base/services/wlansvc/rpcserver.c (hunk 2 above)
Comment (ColinFinck) [reply base/services/wlansvc/rpcserver.c]: This PR is unfinished and waiting to be picked up by someone else (see the TODO list).
Christoph couldn't figure out the crash at Hackfest, so he committed what he had and hopes for someone to find out why the function still crashes at RPC return.

**Comment by KyleKatarn31** [comment]
Please note that the "signature" of the problem has changed (see update on https://jira.reactos.org/browse/CORE-12275 )

**Comment by Saibamen** [comment]
Please mark this Pull Request as Draft if is still WIP: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request

**Comment by GeoB99** [comment]
Any updates on this?

**Comment by ColinFinck** [comment]
> Any updates on this?

Not that I'm aware of. What I said in https://github.com/reactos/reactos/pull/770#discussion_r218548667 basically still applies.

I recall that Christoph figured out the WLAN RPC interface at Hackfest 2018 (using some RPC communication sniffer) and attempted to implement the first parts in this PR. Unfortunately, it remained unfinished in the given time and still crashes at unmarshalling.
I also recall that he wasn't sure about all details of marshalling/unmarshalling structures for the WLAN RPC protocol - for example what to report as size when the caller asks, how strings are returned or how pointers inside returned structures are handled.

These days, I'm pretty confident that the marshalling/unmarshalling has to be done similarly to what I've done for the Printing RPC protocol:
- https://github.com/reactos/reactos/blob/3fa57b8ff7fcee47b8e2ed869aecaf4515603f3f/win32ss/printing/base/marshalling.c
- https://github.com/reactos/reactos/blob/081ffbeb4f02f000aa6aedd9177bf81b62684716/win32ss/printing/base/spoolss/tools.c

All in all, the work left on wlanapi and wlansvc is more than this PR. The components require a maintainer who takes over this PR and also wants to invest some time into them afterwards.
@christophvw can probably tell what he used to figure out the RPC protocol. And I would offer to talk about the Printing RPC protocol if any further information is required. But I guess both of us lack the time to take over this PR.

**Pull request closed without merging**